kfs-core/src/main/java/edu/arizona/kfs/fp/document/DisbursementVoucherDocument.java

```diff
@@ -1,27 +1,205 @@
 
     @Override
     public void prepareForSave() {
         LOG.debug("DisbursementVoucherDocument.prepareForSave()");
         super.prepareForSave();
 
+        DocumentHelperService documentHelperService = SpringContext.getBean(DocumentHelperService.class);
+        DocumentAuthorizer docAuth = documentHelperService.getDocumentAuthorizer(this);
+
+        // First, only do this if the document is in initiated status - after that, we don't want to 
+        // accidentally reset the bank code
+        if ( KewApiConstants.ROUTE_HEADER_INITIATED_CD.equals( getDocumentHeader().getWorkflowDocument().getStatus().getCode() )
+                || KewApiConstants.ROUTE_HEADER_SAVED_CD.equals( getDocumentHeader().getWorkflowDocument().getStatus().getCode() ) ) {
+            // need to check whether the user has the permission to edit the bank code
+            // if so, don't synchronize since we can't tell whether the value coming in
+            // was entered by the user or not.        
+            if ( !docAuth.isAuthorizedByTemplate(this, 
+                    KFSConstants.ParameterNamespaces.KFS, 
+                    KFSConstants.PermissionTemplate.EDIT_BANK_CODE.name, 
+                    GlobalVariables.getUserSession().getPrincipalId()  ) ) {
+                synchronizeBankCodeWithPaymentMethod();        
+            } else {
+                refreshReferenceObject( "bank" );
+            }
+        } 
+        else{           
+            TransactionalDocumentPresentationController presentationController = (TransactionalDocumentPresentationController) documentHelperService.getDocumentPresentationController(this);
+            if(presentationController.getEditModes(this).contains(KFSConstants.Authorization.PAYMENT_METHOD_EDIT_MODE)){
+                synchronizeBankCodeWithPaymentMethod();
+            }
+        }
+        
+        // need to initialize extension primary key values because OJB doesn't do a good job 
         for (Object o : getSourceAccountingLines()) {
             SourceAccountingLine accountingLine = (SourceAccountingLine) o;
             DisbursementVoucherSourceAccountingLineExtension accountingLineExtension = (DisbursementVoucherSourceAccountingLineExtension) accountingLine.getExtension();
             accountingLineExtension.setDocumentNumber(accountingLine.getDocumentNumber());
             accountingLineExtension.setSequenceNumber(accountingLine.getSequenceNumber());
         }
     }
+    
+    protected void synchronizeBankCodeWithPaymentMethod() {
+        Bank bank = getPaymentMethodGeneralLedgerPendingEntryService().getBankForPaymentMethod( getDisbVchrPaymentMethodCode() );
+        if ( bank != null ) {
+            if ( !StringUtils.equals(bank.getBankCode(), getDisbVchrBankCode()) ) {
+                setDisbVchrBankCode(bank.getBankCode());
+                refreshReferenceObject( "bank" );

```

Comment:
Since this is being used more than once, "bank" should be made a class constant.